### ISSUE TITLE
Certain types of invalid HTML tags are parsed as valid HTML

### ISSUE THREAD

**Issue by omninonsense** [YorickPeterse/oga#132]
Seems like `Oga::parse_html` treats certain invalid HTML as valid HTML.
~~~ruby
Oga::parse_html('<3').to_xml
# <3></3>
~~~

Any following text is treated as attirbutes, and any opening tags (`<`) are ignored until one of the closing tags occur (`>` or `/>`) :

~~~ruby
Oga::parse_html('<3 is beautiful').to_xml
# <3 is="" beautiful=""></3>

Oga::parse_html('<3 is beautiful>, but rare').to_xml
# <3 is="" beautiful="">, but rare</3>

Oga::parse_html('<3 is beautiful/>, but rare').to_xml
# <3 is="" beautiful=""></3>, but rare
~~~

Expected result (to be compliant with browsers):

~~~ruby
Oga::parse_html('<3 is beautiful').to_xml
# &lt;3 is beautiful
~~~

Workaround:

~~~ruby
Oga::parse_html('&lt;3 is beautiful').to_xml
# &lt;3 is beautiful
~~~

IIRC (I could be wrong) WebKit handles this by going back after a `<` token isn't matched by a `>` token, and another `<` token appears. When it goes back, it converts everything between the two `<` tokens into a text node, and resumes parsing right before the second `<`.

**Comment by YorickPeterse** [comment]
As I briefly mentioned on IRC, I don't think there's currently a way around
this. The way the XML/HTML lexer works is that the moment it finds a `<`
followed by 1 or more non-whitespace characters it will treat this as an element
opening tag. Once the identifier has been processed the lexer jumps to a
separate state that's used for processing attributes. The lexer will then jump
into the element's "text" state the moment it finds a `>`.

For input such as `<3 is beautiful` this means the lexer will never leave the
state (technically there are a few states, but those are details) used for
processing attributes. For example, if we use the HTML lexer directly we'll get
the following tokens:

    Oga::XML::Lexer.new('<3 is beautiful', :html => true).lex
    [[:T_ELEM_NAME, "3", 1], [:T_ATTR, "is", 1], [:T_ATTR, "beautiful", 1], [:T_ELEM_END, nil, 1]]

Here `T_ELEM_END` is inserted automatically once lexing finishes, hence the
above input doesn't trigger any parsing errors.

There are two possible ways of solving this particular problem, both equally
problematic:

1. When the state for processing attributes is entered, never left explicitly
   (that is, no closing `>` is to be found anywhere) _and_ we've run out of
   input (or we encounter a new opening tag), then the input is lexed as regular
   text.
2. Some form of lookahead would have to be implemented so the lexer is aware of
   tags being closed properly or not. This is extremely difficult to implement,
   especially when supporting streaming of input.

Option 2 is pretty much a no-go. Option 1 is also hard to implement as tokens
have already been emitted once the end-of-input is reached.

In short: there's currently no proper way of fixing this that I can think of at
this time. I'll keep this issue open for some time, maybe I (or somebody else)
comes up with a decent solution.

**Comment by abotalov** [comment]
The rules are actually different as the following series of examples shows:

Source | Result
--- | ---
<3 | &lt;3
<unknown | 

In [Tag open state](http://www.w3.org/TR/html51/syntax.html#tag-open-state) only letters can appear. So `<3` immediately leads to emitting `&lt;`.

**Comment by abotalov** [comment]
In XML, though, numbers can appear in tag names - http://www.w3.org/TR/REC-xml/#NT-NameChar

**Comment by abotalov** [comment]
In HTML the tag token is emitted only when `>` character occurs - http://www.w3.org/TR/html51/syntax.html#tag-name-state / http://www.w3.org/TR/html51/syntax.html#after-attribute-value-(quoted)-state

At the end of the document the nodes are just popped - http://www.w3.org/TR/html51/syntax.html#the-end

**Comment by YorickPeterse** [comment]
I'm going to close this for the time being. As mentioned in https://github.com/YorickPeterse/oga/issues/132#issuecomment-129629987 there's currently no decent way to handle input like this (that I'm aware of) without tearing my hairs out.
